File: 25.k-个一组翻转链表.go
/*
 * @lc app=leetcode.cn id=25 lang=golang
 *
 * [25] K 个一组翻转链表
 */
package main

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup1(head *ListNode, k int) *ListNode {
	if k <= 1 {
		return head
	}
	if !compareKGroup(head, k) {
		return head
	}
	for i := 0; i < k-1; i++ {
		head = swapPairsInter(head, k-i)
	}
	theKthListNode := getKthListNode(head, k)
	theKthListNode.Next = reverseKGroup1(theKthListNode.Next, k)
	return head
}

// 交换的入口
func swapPairsInter(head *ListNode, k int) *ListNode {
	if k <= 1 {
		return head
	}
	if !compareKGroup(head, k) {
		return head
	}
	head = swapPairsListNode(head)
	head.Next = swapPairsInter(head.Next, k-1)
	return head
}

// 交换相邻节点
func swapPairsListNode(leftListNode *ListNode) *ListNode {
	rightListNode := leftListNode.Next
	if leftListNode == nil || rightListNode == nil {
		return leftListNode
	}
	tempListNode1 := leftListNode
	tempListNode2 := rightListNode.Next
	leftListNode = rightListNode
	rightListNode = tempListNode1
	leftListNode.Next = rightListNode
	rightListNode.Next = tempListNode2
	return leftListNode
}

// 判断当前剩余节点数量是否满足K
func compareKGroup(head *ListNode, k int) bool {
	tempListNode := head
	for {
		if tempListNode != nil {
			k--
			if k <= 0 {
				return true
			}
			tempListNode = tempListNode.Next
		} else {
			return false
		}
	}
}

// 得到交换后的最后一个节点
func getKthListNode(head *ListNode, k int) *ListNode {
	tempListNode := head
	for i := 0; i < k-1; i++ {
		tempListNode = tempListNode.Next
	}
	return tempListNode
}

// @lc code=end

File: 25.k-个一组翻转链表.cs
using System.Linq.Expressions;
using System.Globalization;
using System.Collections.Generic;
/*
 * @lc app=leetcode.cn id=25 lang=csharp
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public ListNode ReverseKGroup(ListNode head, int k)
    {
        ListNode hair = new ListNode() { next = head };
        ListNode pre = hair;
        while (head != null)
        {
            ListNode tail = pre;
            for (int i = 0; i < k; i++)
            {
                tail = tail.next;
                if (tail == null)
                {
                    return hair.next;
                }
            }
            ListNode nex = tail.next;
            ListNode[] temp = MyReverse(head,tail);
            head = temp[0];
            tail = temp[1];
            pre.next = head;
            tail.next =nex;
            pre = tail;
            head = tail.next;
        }
        return hair.next;
    }

    public ListNode[] MyReverse(ListNode head, ListNode tail)
    {
        ListNode prev = tail.next;
        ListNode p = head;
        while (prev != tail)
        {
            ListNode nex = p.next;
            p.next = prev;
            prev = p;
            p=nex;
        }
        return new ListNode[]{tail,head};
    }
}
// @lc code=end

